File: Ucp/UcpRtoEstimator.cs
using System;

namespace Ucp
{

    internal sealed class UcpRtoEstimator
    {

        private readonly long _minRtoMicros;

        private readonly long _maxRtoMicros;

        private readonly double _backoffFactor;

        public long SmoothedRttMicros { get; private set; }

        public long RttVarianceMicros { get; private set; }

        public long CurrentRtoMicros { get; private set; }

        public UcpRtoEstimator()
            : this(new UcpConfiguration())
        {
        }

        public UcpRtoEstimator(UcpConfiguration config)
        {
            config = config ?? new UcpConfiguration();
            _minRtoMicros = config.EffectiveMinRtoMicros;
            _maxRtoMicros = config.EffectiveMaxRtoMicros;
            _backoffFactor = config.EffectiveRetransmitBackoffFactor;
            CurrentRtoMicros = Math.Max(_minRtoMicros, UcpConstants.INITIAL_RTO_MICROS);
        }

        public void Update(long sampleMicros)
        {
            if (sampleMicros <= 0)
            {
                return;
            }

            if (0 == SmoothedRttMicros)
            {

                SmoothedRttMicros = sampleMicros;
                RttVarianceMicros = sampleMicros / UcpConstants.UCP_INITIAL_RTTVAR_DIVISOR;
            }
            else
            {

                long delta = SmoothedRttMicros > sampleMicros ? SmoothedRttMicros - sampleMicros : sampleMicros - SmoothedRttMicros;
                RttVarianceMicros = ((RttVarianceMicros * UcpConstants.RTT_VAR_PREVIOUS_WEIGHT) + delta) / UcpConstants.RTT_VAR_DENOM;
                SmoothedRttMicros = ((SmoothedRttMicros * UcpConstants.RTT_SMOOTHING_PREVIOUS_WEIGHT) + sampleMicros) / UcpConstants.RTT_SMOOTHING_DENOM;
            }

            long candidate = SmoothedRttMicros + (UcpConstants.RTO_GAIN_MULTIPLIER * RttVarianceMicros);
            if (candidate < _minRtoMicros)
            {
                candidate = _minRtoMicros;
            }

            if (candidate > _maxRtoMicros)
            {
                candidate = _maxRtoMicros;
            }

            CurrentRtoMicros = candidate;
        }

        public void Backoff()
        {
            double backedOff = CurrentRtoMicros * _backoffFactor;
            double minBackoff = Math.Max(CurrentRtoMicros, _minRtoMicros * UcpConstants.RTO_MAX_BACKOFF_MIN_RTO_MULTIPLIER);
            if (backedOff < minBackoff)
            {
                backedOff = minBackoff;
            }

            if (backedOff > _maxRtoMicros)
            {
                backedOff = _maxRtoMicros;
            }

            CurrentRtoMicros = (long)backedOff;
        }
    }
}

File: cpp/src/ucp_rto_estimator.cpp
/** @file ucp_rto_estimator.cpp
 *  @brief Jacobson/Karels RTO estimator implementation -- mirrors C# Ucp.Internal.RtoEstimator.
 *
 *  Tracks SRTT and RTTVAR via EWMA and computes RTO = SRTT + 4*RTTVAR
 *  (RFC 6298).  Backoff() multiplies the RTO by the configured backoff
 *  factor (default 1.2) on timeout events, floored by configured min/max
 *  bounds and a maximum backoff multiplier.
 */

#include "ucp/ucp_rto_estimator.h"

#include "ucp/ucp_constants.h"

#include <algorithm>

#include <cstdlib>

namespace ucp {

UcpRtoEstimator::UcpRtoEstimator() noexcept : UcpRtoEstimator(UcpConfiguration{}) {}

UcpRtoEstimator::UcpRtoEstimator(const UcpConfiguration& config) noexcept
    : m_min_rto_micros(config.EffectiveMinRtoMicros()), m_max_rto_micros(config.EffectiveMaxRtoMicros()),
      m_backoff_factor(config.EffectiveRetransmitBackoffFactor()) {
    m_current_rto_micros = std::max(m_min_rto_micros, Constants::INITIAL_RTO_MICROS);
}

void UcpRtoEstimator::Update(int64_t sample_micros) noexcept {
    if (0 >= sample_micros) {
        return;
    }

    if (0 == m_srtt_micros) {
        m_srtt_micros = sample_micros;
        m_rttvar_micros = sample_micros / Constants::INITIAL_RTTVAR_DIVISOR;
    } else {
        int64_t delta = std::llabs(m_srtt_micros - sample_micros);

        m_rttvar_micros = ((m_rttvar_micros * Constants::RTT_VAR_PREVIOUS_WEIGHT) + delta) / Constants::RTT_VAR_DENOM;

        m_srtt_micros = ((m_srtt_micros * Constants::RTT_SMOOTHING_PREVIOUS_WEIGHT) + sample_micros) / Constants::RTT_SMOOTHING_DENOM;
    }

    int64_t candidate = m_srtt_micros + (Constants::RTO_GAIN_MULTIPLIER * m_rttvar_micros);

    if (candidate < m_min_rto_micros) {
        candidate = m_min_rto_micros;
    }

    if (candidate > m_max_rto_micros) {
        candidate = m_max_rto_micros;
    }

    m_current_rto_micros = candidate;
}

void UcpRtoEstimator::Backoff() noexcept {
    double backed_off = static_cast<double>(m_current_rto_micros) * m_backoff_factor;

    double min_backoff = std::max(static_cast<double>(m_current_rto_micros),
                                  static_cast<double>(m_min_rto_micros * Constants::RTO_MAX_BACKOFF_MIN_RTO_MULTIPLIER));

    if (backed_off < min_backoff) {
        backed_off = min_backoff;
    }

    if (backed_off > static_cast<double>(m_max_rto_micros)) {
        backed_off = static_cast<double>(m_max_rto_micros);
    }

    m_current_rto_micros = static_cast<int64_t>(backed_off);
}

} // namespace ucp
